src/coreclr/jit/flowgraph.cpp

```diff
@@ -52,6 +52,244 @@ static bool blockNeedsGCPoll(BasicBlock* block)
+                    {
+                        continue;
+                    }
+
+                    GenTree* moduleIdArg = call->fgArgInfo->GetArgNode(0);
+                    GenTree* clsIdArg    = call->fgArgInfo->GetArgNode(1);
+
+                    if (!moduleIdArg->IsCnsIntOrI() || !clsIdArg->IsCnsIntOrI())
+                    {
+                        // Looks like moduleId or/and clsId were passed as indirect loads
+                        continue;
+                    }
+
+                    bool hasNotSupportedPreds = false;
+                    for (flowList* pred = block->bbPreds; pred != nullptr; pred = pred->flNext)
+                    {
+                        BasicBlock* predBlock = pred->getBlock();
+                        if ((predBlock->bbJumpKind != BBJ_NONE) && (predBlock->bbJumpKind != BBJ_ALWAYS) &&
+                            (predBlock->bbJumpKind != BBJ_COND))
+                        {
+                            hasNotSupportedPreds = true;
+                            break;
+                        }
+                        if (!BasicBlock::sameEHRegion(predBlock, block))
+                        {
+                            hasNotSupportedPreds = true;
+                            break;
+                        }
+                    }
+                    if (hasNotSupportedPreds)
+                    {
+                        continue;
+                    }
+
+                    if (prevBb == nullptr)
+                    {
+                        // We're going to emit a BB in front of fgFirstBB
+                        fgEnsureFirstBBisScratch();
+                        prevBb = fgFirstBB;
+                        if (prevBb == block)
+                        {
+                            continue;
+                        }
+                    }
+
+                    // So, we found a helper call inside the "block" - let's extract it to a
+                    // separate block "callInitBb" and guard it with a fast "isInitedBb" bb.
+                    // The final layout should look like this:
+
+                    // BB0 "prevBb":
+                    //     ...
+                    //
+                    // BB1 "isInitedBb":    (preds: BB0 + %current preds of BB3%)
+                    //
+                    //  *  JTRUE     void
+                    //  \--*  NE        int
+                    //     +--*  AND       int
+                    //     |  +--*  IND       ubyte
+                    //     |  |  \--*  CNS_INT   long   moduleIdArg + clsIdArg + dataBlobOffset
+                    //     |  \--*  CNS_INT   int    isInitMask
+                    //     \--*  CNS_INT   int    0
+                    //
+                    //
+                    // BB2 "callInitBb":    (preds: BB1)
+                    //
+                    //  *  CALL help long   HELPER.CORINFO_HELP_GETSHARED_NONGCSTATIC_BASE
```

Comment:
Is this the only helper that qualifies? 

Presumably the other helpers that trigger class init don't allow simple tests for class init, so the checking is more involved and not worth the trouble? Probably worth a comment.

Reply:
I assume there are more can be done, Jan mentioned DYNAMICCLASS. Also, I guess, we also optimze R2R with indirect loads.

Also, what about ThreadStatic - if we had a register where we could always get current thread context we could access thread-static fields faster.